### ISSUE TITLE
feature request: Support ItemBatcher in StepFunctions

### ISSUE THREAD

**Issue by noseworthy** [localstack/localstack#10884]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature description

We'd like to optimize our Step Function State Machine Distributed Map execution by using the [ItemBatcher](https://docs.aws.amazon.com/step-functions/latest/dg/input-output-itembatcher.html) syntax. I've built out a `cdk` app to deploy a StateMachine using `ItemBatcher` and I get `ASLParserException`s. I'm assuming that means that `localstack` doesn't support `ItemBatcher` yet as I can't find any reference to it in the repository.

```
❌ Deployment failed: Error: The stack named XXXXXXXXXXXXX failed to deploy: CREATE_FAILED (An error occurred (InvalidDefinition) when calling the CreateStateMachine operation: ASLParserException ['line 1:3169, at "ItemBatcher", mismatched input \'"ItemBatcher"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3206, at "BatchInput", mismatched input \'"BatchInput"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3258, at "documentStoreSyncId", mismatched input \'"documentStoreSyncId"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3304, at "documentId", mismatched input \'"documentId"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3335, at "extract-clauses-task", mismatched input \'"extract-clauses-task"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}']): An error occurred (InvalidDefinition) when calling the CreateStateMachine operation: ASLParserException ['line 1:3169, at "ItemBatcher", mismatched input \'"ItemBatcher"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3206, at "BatchInput", mismatched input \'"BatchInput"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3258, at "documentStoreSyncId", mismatched input \'"documentStoreSyncId"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3304, at "documentId", mismatched input \'"documentId"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}', 'line 1:3335, at "extract-clauses-task", mismatched input \'"extract-clauses-task"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"ResultSelector"\', \'"ItemReader"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"Error"\', \'"Retry"\', \'"Catch"\'}']
```

### Anything else?

- This is a new ticket "reopening" https://github.com/localstack/localstack/issues/9758 which is still an issue in v3.4.0 of localstack-pro
- `ItemBatcher` syntax for ASL - https://docs.aws.amazon.com/step-functions/latest/dg/input-output-itembatcher.html

**Comment by imsagark** [comment]
Has support for ItemBatcher in Step Functions been added to LocalStack in recent updates?

**Comment by dbryantm** [comment]
I'm also running into issues with `ResultWriter` and `Transformation`. I get errors like the following: 

```
╷
│ Error: invalid Step Functions State Machine definition: ERROR (SCHEMA_VALIDATION_FAILED): ASLParserException ['line 74:12, at "ItemBatcher", mismatched input \'"ItemBatcher"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrencyPath"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"Items"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"Credentials"\', \'"ResultSelector"\', \'"ItemReader"\', \'"ToleratedFailureCount"\', \'"ToleratedFailureCountPath"\', \'"ToleratedFailurePercentage"\', \'"ToleratedFailurePercentagePath"\', \'"Label"\', \'"ResultWriter"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"CausePath"\', \'"Error"\', \'"ErrorPath"\', \'"Retry"\', \'"Catch"\', \'"QueryLanguage"\', \'"Assign"\', \'"Output"\', \'"Arguments"\'}', 'line 76:14, at "BatchInput", mismatched input \'"BatchInput"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrencyPath"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"Items"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"Credentials"\', \'"ResultSelector"\', \'"ItemReader"\', \'"ToleratedFailureCount"\', \'"ToleratedFailureCountPath"\', \'"ToleratedFailurePercentage"\', \'"ToleratedFailurePercentagePath"\', \'"Label"\', \'"ResultWriter"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"CausePath"\', \'"Error"\', \'"ErrorPath"\', \'"Retry"\', \'"Catch"\', \'"QueryLanguage"\', \'"Assign"\', \'"Output"\', \'"Arguments"\'}', 'line 78:16, at "sourceFile", mismatched input \'"sourceFile"\' expecting {\'"Comment"\', \'"Type"\', \'"Choices"\', \'"Default"\', \'"Branches"\', \'"SecondsPath"\', \'"Seconds"\', \'"TimestampPath"\', \'"Timestamp"\', \'"TimeoutSeconds"\', \'"TimeoutSecondsPath"\', \'"HeartbeatSeconds"\', \'"HeartbeatSecondsPath"\', \'"ItemProcessor"\', \'"Iterator"\', \'"ItemSelector"\', \'"MaxConcurrencyPath"\', \'"MaxConcurrency"\', \'"Resource"\', \'"InputPath"\', \'"OutputPath"\', \'"Items"\', \'"ItemsPath"\', \'"ResultPath"\', \'"Result"\', \'"Parameters"\', \'"Credentials"\', \'"ResultSelector"\', \'"ItemReader"\', \'"ToleratedFailureCount"\', \'"ToleratedFailureCountPath"\', \'"ToleratedFailurePercentage"\', \'"ToleratedFailurePercentagePath"\', \'"Label"\', \'"ResultWriter"\', \'"Next"\', \'"End"\', \'"Cause"\', \'"CausePath"\', \'"Error"\', \'"ErrorPath"\', \'"Retry"\', \'"Catch"\', \'"QueryLanguage"\', \'"Assign"\', \'"Output"\', \'"Arguments"\'}', 'line 81:12, at "ResultWriter", mismatched input \'"ResultWriter"\' expecting {\'"Comment"\', \'"States"\', \'"StartAt"\', \'"ProcessorConfig"\'}', 'line 84:16, at "Transformation", mismatched input \'"Transformation"\' expecting {\'"Comment"\', \'"States"\', \'"StartAt"\', \'"ProcessorConfig"\'}', "line 87:19, at true, mismatched input 'true' expecting '{'", "line 90:6, at }, mismatched input '}' expecting <EOF>"]
```

Not sure if the output could help in the resolution of this but figured I would pass it along anyway.

**Comment by drauedo** [comment]
Hello guys,

I just wanted to provide a quick update on this issue.

We're currently planning to implement this by the end of October. We'll keep you posted about any updates here.

**Comment by barbarosalp** [comment]
Hi, I am blocked by this at the moment. Any updates? Thanks!

**Comment by tiurin** [comment]
Hi @barbarosalp, thanks for bringing attention to the issue. Unfortunately, the implementation timeline has been pushed further due to priority changes - we've been busy implementing [TestState API enhancements](https://blog.localstack.cloud/effective-unit-testing-for-aws-step-functions/). Current tentative plan is to add support for ItemBatcher by the end of February. We'll post an update once the implementation is actually in progress.

**Comment by adrian-fabric** [comment]
Please add support for ItemBatcher. There are some workarounds, but we they are not as efficient as ItemBatcher.

**Comment by HarshCasper** [comment]
Due to the forthcoming changes in our distribution and development model ([read more](https://blog.localstack.cloud/the-road-ahead-for-localstack/)), this issue tracker is no longer being maintained. We've migrated this issue to our internal tracker and will continue to work on it there. If you'd like to follow up on the status or expedite this, feel free to reach out through one of our support channels:

-   💬 [GitHub Discussions](https://github.com/orgs/localstack/discussions)
-   📖 [Help & Support](https://docs.localstack.cloud/aws/getting-started/help-support/)
